<a href="https://colab.research.google.com/github/paulrobustelli/CHEM76_S26/blob/main/Huckel_Calculations.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [27]:
import numpy as np
from numpy import linalg
from numpy.linalg import *
import scipy
from scipy.linalg import schur, eigvals, svd

In [28]:
a=np.zeros((2,2))

In [29]:
a

array([[0., 0.],
       [0., 0.]])

In [30]:
a[0]

array([0., 0.])

In [31]:
a[0]=[1,0]
a[1]=[0,1]
a

array([[1., 0.],
       [0., 1.]])

In [126]:
a = np.array([[1,2],[3,4]])
b = np.array([[2,4],[6,8]])
print(a)
print(b)

[[1 2]
 [3 4]]
[[2 4]
 [6 8]]


In [127]:
print(a)
det(a)

[[1 2]
 [3 4]]


np.float64(-2.0000000000000004)

In [128]:
print(b)
det(b)

[[2 4]
 [6 8]]


np.float64(-8.000000000000002)

In [129]:
print(a)
print(b)
np.dot(a,b)

[[1 2]
 [3 4]]
[[2 4]
 [6 8]]


array([[14, 20],
       [30, 44]])

In [130]:
c=np.linalg.inv(a)
print(c)

[[-2.   1. ]
 [ 1.5 -0.5]]


In [131]:
np.dot(a,c)

array([[1.0000000e+00, 0.0000000e+00],
       [8.8817842e-16, 1.0000000e+00]])

In [132]:
M=np.array([[-6,3],[4,5]])
M

array([[-6,  3],
       [ 4,  5]])

In [133]:
eigvec1=np.array([1,4])
M1=np.dot(M,eigvec1)
print("M1",M1)
eigvec1*6
#Therefore 6 is the eigenvalue and [1,4] is an eigenvector

M1 [ 6 24]


array([ 6, 24])

# Ethene Hückel Matrix Formulation Summary

## Matrix Method, write $H$

$$
\begin{pmatrix}
\alpha_A & \beta_{AB} \\
\beta_{AB} & \alpha_B
\end{pmatrix}
\begin{pmatrix}
c_A \\
c_B
\end{pmatrix}
=
E
\begin{pmatrix}
c_A \\
c_B
\end{pmatrix}
$$

---

$$
H =
\begin{pmatrix}
\alpha & \beta \\
\beta & \alpha
\end{pmatrix}
$$

$$
H = \alpha I + \beta M
$$

$$
I =
\begin{pmatrix}
1 & 0 \\
0 & 1
\end{pmatrix}
\qquad
M =
\begin{pmatrix}
0 & 1 \\
1 & 0
\end{pmatrix}
$$

---

## Diagonalizing $M$

In [186]:
A=np.array([[1, 0.],
       [0., 1]])
M=np.array([[0, 1.],
       [1., 0]])
print(A)
print(M)
eigvals,C=np.linalg.eig(M)
print(eigvals)
print(C)

[[1. 0.]
 [0. 1.]]
[[0. 1.]
 [1. 0.]]
[ 1. -1.]
[[ 0.70710678 -0.70710678]
 [ 0.70710678  0.70710678]]


$$
D =
\begin{pmatrix}
1 & 0 \\
0 & -1
\end{pmatrix}
$$

$$
D \text{ is the eigenvalue matrix}
$$

$$
\text{The diagonal elements of } D \text{ are the eigenvalues of } M
$$

$$
C =
\frac{1}{\sqrt{2}}
\begin{pmatrix}
1 & -1 \\
1 & 1
\end{pmatrix}
$$

$$
\text{The columns of } C \text{ are the eigenvectors of } M
$$

$$
C^{-1} M C = D
$$

$$
C^{-1} M C = D
$$


$$
D =
\begin{pmatrix}
1 & 0 \\
0 & -1
\end{pmatrix}
$$


In [188]:
D=np.dot(np.dot(np.linalg.inv(C),M),C)
np.around(D,decimals=3)

array([[ 1., -0.],
       [ 0., -1.]])

## Construct $E$ from $\alpha I$ and $D$

$$
H = \alpha I + \beta M \;\rightarrow\; E = \alpha I + \beta D
$$

$$
E =
\begin{pmatrix}
\alpha & 0 \\
0 & \alpha
\end{pmatrix}
+
\begin{pmatrix}
\beta & 0 \\
0 & -\beta
\end{pmatrix}
$$

$$
E =
\begin{pmatrix}
\alpha + \beta & 0 \\
0 & \alpha - \beta
\end{pmatrix}
$$

---

## Construct molecular orbitals from $C$

$$
C =
\frac{1}{\sqrt{2}}
\begin{pmatrix}
1 & -1 \\
1 & 1
\end{pmatrix}
$$

$$
\text{Columns of } C \text{ give the coefficients of the molecular orbitals}
$$

$$
\psi_1 = c_{A1}\psi_A + c_{B1}\psi_B
\qquad
\psi_2 = c_{A2}\psi_A + c_{B2}\psi_B
$$

---

$$
\psi_1 =
\frac{1}{\sqrt{2}}
(1\psi_A + 1\psi_B)
=
\frac{1}{\sqrt{2}}(\psi_A + \psi_B)
$$

$$
\psi_2 =
\frac{1}{\sqrt{2}}
(-1\psi_A + 1\psi_B)
=
\frac{1}{\sqrt{2}}(\psi_B - \psi_A)
$$

---

$$
E_+ = \alpha + \beta
\qquad
\psi_1 = \frac{1}{\sqrt{2}}(\psi_A + \psi_B)
$$

$$
E_- = \alpha - \beta
\qquad
\psi_2 = \frac{1}{\sqrt{2}}(\psi_A - \psi_B)
$$

## Butadiene Hückel Matrix

For butadiene, there are four neighboring \(p_z\) orbitals:

$$
\psi_A,\;\psi_B,\;\psi_C,\;\psi_D
$$

The Hückel Hamiltonian is

$$
H =
\begin{pmatrix}
\alpha & \beta & 0 & 0 \\
\beta & \alpha & \beta & 0 \\
0 & \beta & \alpha & \beta \\
0 & 0 & \beta & \alpha
\end{pmatrix}
$$

We can write this as

$$
H = \alpha I + \beta M
$$

where

$$
I =
\begin{pmatrix}
1 & 0 & 0 & 0 \\
0 & 1 & 0 & 0 \\
0 & 0 & 1 & 0 \\
0 & 0 & 0 & 1
\end{pmatrix}
$$

and

$$
M =
\begin{pmatrix}
0 & 1 & 0 & 0 \\
1 & 0 & 1 & 0 \\
0 & 1 & 0 & 1 \\
0 & 0 & 1 & 0
\end{pmatrix}
$$

---

## Diagonalize \(M\)

In [197]:
M = np.zeros((4,4))

M[0] = [0,1,0,0]
M[1] = [1,0,1,0]
M[2] = [0,1,0,1]
M[3] = [0,0,1,0]


A = np.zeros((4,4))

A[0] = [1,0,0,0]
A[1] = [0,1,0,0]
A[2] = [0,0,1,0]
A[3] = [0,0,0,1]


print(A)
print(M)
eigvals,C=np.linalg.eig(M)
print("Eigenvalues:")
print(eigvals)
print("Coefficent Matrix")
print(C)

[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]
[[0. 1. 0. 0.]
 [1. 0. 1. 0.]
 [0. 1. 0. 1.]
 [0. 0. 1. 0.]]
Eigenvalues:
[-1.61803399 -0.61803399  1.61803399  0.61803399]
Coefficent Matrix
[[ 0.37174803  0.60150096 -0.37174803 -0.60150096]
 [-0.60150096 -0.37174803 -0.60150096 -0.37174803]
 [ 0.60150096 -0.37174803 -0.60150096  0.37174803]
 [-0.37174803  0.60150096 -0.37174803  0.60150096]]


In [192]:
np.around(C,decimals=3)

array([[ 0.372,  0.602, -0.372, -0.602],
       [-0.602, -0.372, -0.602, -0.372],
       [ 0.602, -0.372, -0.602,  0.372],
       [-0.372,  0.602, -0.372,  0.602]])

## Diagonalizing $M$

$$
M =
\begin{pmatrix}
0 & 1 & 0 & 0 \\
1 & 0 & 1 & 0 \\
0 & 1 & 0 & 1 \\
0 & 0 & 1 & 0
\end{pmatrix}
$$

The NumPy eigenvalues are returned in this order:

$$
\lambda = -1.618,\;-0.618,\;1.618,\;0.618
$$

---

## Eigenvector matrix $C$

$$
C =
\begin{pmatrix}
0.372 & 0.602 & -0.372 & -0.602 \\
-0.602 & -0.372 & -0.602 & -0.372 \\
0.602 & -0.372 & -0.602 & 0.372 \\
-0.372 & 0.602 & -0.372 & 0.602
\end{pmatrix}
$$

$$
\text{Columns of } C \text{ are the eigenvectors of } M
$$

---

$$
C^{-1}MC = D
$$

$$
D =
\begin{pmatrix}
-1.618 & 0 & 0 & 0 \\
0 & -0.618 & 0 & 0 \\
0 & 0 & 1.618 & 0 \\
0 & 0 & 0 & 0.618
\end{pmatrix}
$$

---

## Construct $E$ from $\alpha I$ and $D$

Since

$$
H = \alpha I + \beta M
$$

diagonalizing $M$ gives

$$
E = \alpha I + \beta D
$$

Therefore,

$$
E =
\begin{pmatrix}
\alpha - 1.618\beta & 0 & 0 & 0 \\
0 & \alpha - 0.618\beta & 0 & 0 \\
0 & 0 & \alpha + 1.618\beta & 0 \\
0 & 0 & 0 & \alpha + 0.618\beta
\end{pmatrix}
$$

---

## Molecular orbitals from $C$

Each column of $C$ gives the coefficients for one molecular orbital.

$$
\psi_1 =
0.372\psi_A - 0.602\psi_B + 0.602\psi_C - 0.372\psi_D
$$

$$
\psi_2 =
0.602\psi_A - 0.372\psi_B - 0.372\psi_C + 0.602\psi_D
$$

$$
\psi_3 =
-0.372\psi_A - 0.602\psi_B - 0.602\psi_C - 0.372\psi_D
$$

$$
\psi_4 =
-0.602\psi_A - 0.372\psi_B + 0.372\psi_C + 0.602\psi_D
$$

---

## Energies and molecular orbitals

$$
E_1 = \alpha - 1.618\beta
\qquad
\psi_1 =
0.372\psi_A - 0.602\psi_B + 0.602\psi_C - 0.372\psi_D
$$

$$
E_2 = \alpha - 0.618\beta
\qquad
\psi_2 =
0.602\psi_A - 0.372\psi_B - 0.372\psi_C + 0.602\psi_D
$$

$$
E_3 = \alpha + 1.618\beta
\qquad
\psi_3 =
-0.372\psi_A - 0.602\psi_B - 0.602\psi_C - 0.372\psi_D
$$

$$
E_4 = \alpha + 0.618\beta
\qquad
\psi_4 =
-0.602\psi_A - 0.372\psi_B + 0.372\psi_C + 0.602\psi_D
$$

## Cyclobutadiene Hückel Matrix

For cyclobutadiene, each carbon is bonded to two neighboring carbons in a ring.

$$
H = \alpha I + \beta M
$$

where

$$
M =
\begin{pmatrix}
0 & 1 & 0 & 1 \\
1 & 0 & 1 & 0 \\
0 & 1 & 0 & 1 \\
1 & 0 & 1 & 0
\end{pmatrix}
$$

In [199]:
import numpy as np

# Cyclobutadiene M matrix
M = np.zeros((4,4))

M[0] = [0,1,0,1]
M[1] = [1,0,1,0]
M[2] = [0,1,0,1]
M[3] = [1,0,1,0]

print(M)

[[0. 1. 0. 1.]
 [1. 0. 1. 0.]
 [0. 1. 0. 1.]
 [1. 0. 1. 0.]]


## Diagonalize $M$

Now we diagonalize $M$ to obtain the eigenvalues and coefficient matrix $C$.

The columns of $C$ are the eigenvectors of $M$.

In [200]:
eigvals, C = np.linalg.eig(M)

print("Eigenvalues:")
print(eigvals)

print("Coefficient Matrix:")
print(C)

print("Rounded Coefficient Matrix:")
print(np.round(C, decimals=3))

Eigenvalues:
[-2.00000000e+00  6.59737022e-17  2.00000000e+00  0.00000000e+00]
Coefficient Matrix:
[[ 5.00000000e-01 -7.07106781e-01  5.00000000e-01  0.00000000e+00]
 [-5.00000000e-01 -4.63633825e-17  5.00000000e-01 -7.07106781e-01]
 [ 5.00000000e-01  7.07106781e-01  5.00000000e-01  0.00000000e+00]
 [-5.00000000e-01 -3.76905605e-17  5.00000000e-01  7.07106781e-01]]
Rounded Coefficient Matrix:
[[ 0.5   -0.707  0.5    0.   ]
 [-0.5   -0.     0.5   -0.707]
 [ 0.5    0.707  0.5    0.   ]
 [-0.5   -0.     0.5    0.707]]


## Eigenvalue matrix $D$

The eigenvalues are

$$
\lambda = -2,\;0,\;2,\;0
$$

so the diagonal eigenvalue matrix is

$$
D =
\begin{pmatrix}
-2 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 \\
0 & 0 & 2 & 0 \\
0 & 0 & 0 & 0
\end{pmatrix}
$$

The matrix $C$ diagonalizes $M$:

$$
C^{-1}MC = D
$$

In [201]:
D = np.dot(np.dot(np.linalg.inv(C), M), C)
np.round(D, decimals=3)

array([[-2.,  0., -0., -0.],
       [ 0., -0.,  0.,  0.],
       [ 0., -0.,  2.,  0.],
       [-0.,  0., -0., -0.]])

## Construct $E$ from $\alpha I$ and $D$

Since

$$
H = \alpha I + \beta M
$$

diagonalizing $M$ gives

$$
E = \alpha I + \beta D
$$

Therefore,

$$
E =
\begin{pmatrix}
\alpha - 2\beta & 0 & 0 & 0 \\
0 & \alpha & 0 & 0 \\
0 & 0 & \alpha + 2\beta & 0 \\
0 & 0 & 0 & \alpha
\end{pmatrix}
$$

In [204]:
#Benzene:
M=np.zeros((6,6))
M[0]=[0,1,0,0,0,1]
M[1]=[1,0,1,0,0,0]
M[2]=[0,1,0,1,0,0]
M[3]=[0,0,1,0,1,0]
M[4]=[0,0,0,1,0,1]
M[5]=[1,0,0,0,1,0]

eigvals, C = np.linalg.eig(M)

print("Eigenvalues:")
print(eigvals)

print("Coefficient Matrix:")
print(C)

print("Rounded Coefficient Matrix:")
print(np.round(C, decimals=3))

Eigenvalues:
[-2.  2.  1. -1. -1.  1.]
Coefficient Matrix:
[[ 0.40824829 -0.40824829 -0.57735027  0.57732307  0.06604706  0.09791921]
 [-0.40824829 -0.40824829 -0.28867513 -0.29351503 -0.5297411  -0.4437968 ]
 [ 0.40824829 -0.40824829  0.28867513 -0.28380804  0.46369403 -0.54171601]
 [-0.40824829 -0.40824829  0.57735027  0.57732307  0.06604706 -0.09791921]
 [ 0.40824829 -0.40824829  0.28867513 -0.29351503 -0.5297411   0.4437968 ]
 [-0.40824829 -0.40824829 -0.28867513 -0.28380804  0.46369403  0.54171601]]
Rounded Coefficient Matrix:
[[ 0.408 -0.408 -0.577  0.577  0.066  0.098]
 [-0.408 -0.408 -0.289 -0.294 -0.53  -0.444]
 [ 0.408 -0.408  0.289 -0.284  0.464 -0.542]
 [-0.408 -0.408  0.577  0.577  0.066 -0.098]
 [ 0.408 -0.408  0.289 -0.294 -0.53   0.444]
 [-0.408 -0.408 -0.289 -0.284  0.464  0.542]]


## Molecular orbitals for benzene

Each column of $C$ gives the coefficients for one molecular orbital.

$$
\psi_1 =
0.408\psi_A - 0.408\psi_B + 0.408\psi_C - 0.408\psi_D + 0.408\psi_E - 0.408\psi_F
$$

$$
\psi_2 =
-0.408\psi_A - 0.408\psi_B - 0.408\psi_C - 0.408\psi_D - 0.408\psi_E - 0.408\psi_F
$$

$$
\psi_3 =
-0.577\psi_A - 0.289\psi_B + 0.289\psi_C + 0.577\psi_D + 0.289\psi_E - 0.289\psi_F
$$

$$
\psi_4 =
0.577\psi_A - 0.294\psi_B - 0.284\psi_C + 0.577\psi_D - 0.294\psi_E - 0.284\psi_F
$$

$$
\psi_5 =
0.066\psi_A - 0.530\psi_B + 0.464\psi_C + 0.066\psi_D - 0.530\psi_E + 0.464\psi_F
$$

$$
\psi_6 =
0.098\psi_A - 0.444\psi_B - 0.542\psi_C - 0.098\psi_D + 0.444\psi_E + 0.542\psi_F
$$

## Energies and molecular orbitals for benzene

The eigenvalues of $M$ are

$$
\lambda = -2,\; -1,\; -1,\; 1,\; 1,\; 2
$$

so the Hückel energies are

$$
E = \alpha + \beta \lambda
$$

---

$$
E_1 = \alpha - 2\beta
\qquad
\psi_1 =
0.408\psi_A - 0.408\psi_B + 0.408\psi_C - 0.408\psi_D + 0.408\psi_E - 0.408\psi_F
$$

$$
E_2 = \alpha - \beta
\qquad
\psi_2 =
-0.577\psi_A - 0.289\psi_B + 0.289\psi_C + 0.577\psi_D + 0.289\psi_E - 0.289\psi_F
$$

$$
E_3 = \alpha - \beta
\qquad
\psi_3 =
0.577\psi_A - 0.294\psi_B - 0.284\psi_C + 0.577\psi_D - 0.294\psi_E - 0.284\psi_F
$$

---

$$
E_4 = \alpha + \beta
\qquad
\psi_4 =
0.066\psi_A - 0.530\psi_B + 0.464\psi_C + 0.066\psi_D - 0.530\psi_E + 0.464\psi_F
$$

$$
E_5 = \alpha + \beta
\qquad
\psi_5 =
0.098\psi_A - 0.444\psi_B - 0.542\psi_C - 0.098\psi_D + 0.444\psi_E + 0.542\psi_F
$$

$$
E_6 = \alpha + 2\beta
\qquad
\psi_6 =
-0.408\psi_A - 0.408\psi_B - 0.408\psi_C - 0.408\psi_D - 0.408\psi_E - 0.408\psi_F
$$